relay-server/src/services/server/io.rs

```diff
@@ -0,0 +1,265 @@
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use futures::FutureExt;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::time::{Instant, Sleep};
+
+use crate::statsd::RelayCounters;
+
+/// A wrapper for [`AsyncRead`] and [`AsyncWrite`] streams with a maximum idle time.
+///
+/// If there is no activity in the underlying stream for the specified `timeout`
+/// the [`IdleTimeout`] will abort the operation and return [`std::io::ErrorKind::TimedOut`].
+pub struct IdleTimeout<T> {
+    inner: T,
+    timeout: Option<Duration>,
+    // `Box::pin` the sleep timer, the entire connection/stream is required to be `Unpin` anyways.
+    timer: Option<Pin<Box<Sleep>>>,
+    is_idle: bool,
+}
+
+impl<T: Unpin> IdleTimeout<T> {
+    /// Creates a new [`IdleTimeout`] with the specified timeout.
+    ///
+    /// A `None` timeout is equivalent to an infinite timeout.
+    pub fn new(inner: T, timeout: Option<Duration>) -> Self {
+        Self {
+            inner,
+            timeout,
+            timer: None,
+            is_idle: false,
```

Comment:
Would this work with a single variable `deadline: Instant`, i.e. let `Poll::Ready` reset the deadline to a point in the future and `Poll::Pending` check it?

Is the usage of a timer + boolean an optimization to reduce calls to `Instant::now()`?